nixos/dokuwiki: Overhaul for structured settings

e1mo opened the pull request:
###### Description of changes

This PR is a overhaul of the `dokuwiki` module towards the concepts laid out in RFC42. Structured settings can be passed via `services.dokuwiki.sites.<name>.settings`, `.<name>.{aclUse,superUser,disableActions}` have been deprecated accordingly. Both `acl` and `pluginsConfig` now also accept structured settings.
This is related to #209246 and #209253 as part of a larger touch-up of the DokuWiki module, package and tests.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

sinavir commented:
Hello ! This PR looks really really nice, however it doesn't tackle all the problem we can encounter with the dokuwiki packaging for nixos. Especially the test don't load the css from the template (mainly iirc because the bootstrap template needs to source files from dokuwiki (there are some `../` imports somewhere in the source code of the template).

Steps to see it :
```
nix-build -A nixosTests.dokuwiki.driverInteractive
QEMU_NET_OPTS="hostfwd=tcp::8888-:80" ./result/bin/nixos-test-driver
```
Then start one vm in the repl (eg : èdokuwiki_nginx.start()`) and browse `127.0.0.1:8888` on your favorite web browser

It could be really nice if this PR also address this problem but I also understand that it is not trivial. When I thought to it several month ago I went into two possible solutions (with the help of @RaitoBezarius ) :
 - Creating a new derivation with everything packaged together (creating a `withTemplate` function that return the right derivation that vendors everything)
 - Patching the template to get the dokuwiki path at runtime through an environment variable (I believe this is quite dirty)

I also may find the motivation to do it now that the main refactoring work is done

e1mo commented:
Thanks for your feedback!

> It could be really nice if this PR also address this problem but I also understand that it is not trivial. When I thought to it several month ago I went into two possible solutions (with the help of @RaitoBezarius ) :
> 
> * Creating a new derivation with everything packaged together (creating a `withTemplate` function that return the right derivation that vendors everything)
> * Patching the template to get the dokuwiki path at runtime through an environment variable (I believe this is quite dirty)

I have implemented something comparable to your first proposal, but opted to make it a bite more flexible. Using `dokuwiki.combine` it is possible to pass templates and plugins (even tho in my prod deployment all plugins seem to work fine, but that way we can hopefully just prevent issues like these in the future). Without the `basePackage`, any previous overrides (e.g. `dw_with_acronyms`) would not get applied, causing the acronyms test to fail.
I am not sure wehter moving everything dokuwiki-building related into the package definition is correct, but having everything package-building-related in one place seemed sensible to me.

---

Somewhat unrelated since I initially forgot: Pinging @onny and @dali99 as module maintainers for a review.

onny commented:
Thanks for pinging and thank you for working on the Dokuwiki module :) The idea to put the options `superuser`, `useacl` into `dokuwiki.settings` option is a great cleanup :)
Unfortunately the PR seems really big with a lot of changes. Maybe we could split some things up into seperate PRs?
I even would split up the single commit `nixos/dokuwiki: Overhaul for structured settings` into separate PRs for easier review and to get changes faster merged:
- Put change from `mkDerivation` to `overrideAttrs` into a separate PR
- `aclOpts` addition into a separate PR
- Changes to the examples for packaging themes and plugins into a separate PR

As I said, the changes are really cool :) Would really like to help review it

e1mo committed:
nixos/dokuwiki: Overhaul for structured settings

Added the RFC42-style added the posibility to use
`services.dokuwiki.sites.<name>.settings' instead of passing a plain
string to `<name>.extraConfig`. ´<name>.pluginsConfig` now also accepts
structured configuration.

e1mo commented on nixos/modules/services/web-apps/dokuwiki.nix:
```diff
@@ -75,9 +153,22 @@ let
 
         acl = mkOption {
-          type = types.nullOr types.lines;
+          type = with types; nullOr (oneOf [ lines (listOf (submodule aclOpts)) ]);
```
What are the opinions on entirely deprecating support for plaintext ACLs? Since their definition is very straight forward (lines of three, space or tab separated, values) I am confident that `aclOpts` does not impose any limits over the big string.

We could add a warning for everyone passing strings and remove that in time for 23.05.

onny replied:
I'm not using any acls so I cannot tell but sounds fine to me

e1mo commented on nixos/modules/services/web-apps/dokuwiki.nix:
```diff
@@ -100,42 +191,22 @@ let
-            $plugins['authmysql'] = 0;
-            $plugins['authpgsql'] = 0;
-          '';
+          type = with types; oneOf [lines (attrsOf bool)];
```
Same as with `acl`. `attrsOf bool` seems appropriate to cover everything the `plugins.local.php` is intended for.

onny replied:
Sounds good to me

e1mo commented:
> Only the nested configuration line `tpl.mindthedark.autoDark` fails. Not sure how to configure parameters like this with the new `settings` option.

Thanks for pointing that out, that was a small oversight. Is fixed now and the tests now also validate the nested config-generation.

> Will the `extraConfig` option be dropped in the future or is it still required?

Looking at my deployment(s) I would not miss it. In theory one could have settings which e.h. change in regard to the users IP address. But with the config caching employed by Dokuwiki... Not sure how good this would actually work :thinking:
Unless anyone can come p (the more cursed the more interested I am in it) use case which is not covered with the `._raw` helper (e.g. different templates based on Users IP-Address, however this would probably be broken by the config caching?) I'd be fine / in favor of also drop-deprecating this.

onny reviewed (approved):
Works for me and is looking good, thank you :D 
Not sure if you'll incorporate the other changes you mentioned, otherwise we could merge it

e1mo commented:
I need to rebase onto maser anyways, so I'll just throw those deprecation-warnings in while I'm at it.

e1mo commented:
Let's hope I didn't brick anything in between three rebases :sweat_smile:
Oh an one more thing, how would it be in terms of me becoming a co-maintainer of the module and/or tests? After having touched (what at least feels like) a very seizable portion of the module I'd be kinda interested in taking on some sort of "responsibility" (at least for my fresh bugs). I didn't find some process / documentation so I thought i would just ask ^^

onny commented:
Yeah would be nice, of course!

onny commented:
Result of `nixpkgs-review pr 208299` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 test built:</summary>
  <ul>
    <li>nixosTests.dokuwiki</li>
  </ul>
</details>

onny commented:
You could add yourself as a maintainer in a separate PR if you want :)
